radicle/src/cob/issue.rs

```diff
@@ -0,0 +1,766 @@
+        let author = self.store.author();
+        let mut issue = self.store.get_raw(issue_id)?;
+        let changes = events::react(&mut issue, comment_id, &author, &[reaction])?;
+
+        self.store.update(*issue_id, "React", changes, signer)?;
+
+        Ok(())
+    }
+
+    /// Reply to an issue comment.
+    pub fn reply<G: Signer>(
+        &self,
+        issue_id: &IssueId,
+        comment_id: CommentId,
+        reply: &str,
+        signer: &G,
+    ) -> Result<(), Error> {
+        let author = self.store.author();
+        let mut issue = self.store.get_raw(issue_id)?;
+        let changes = events::reply(&mut issue, comment_id, &author, reply, Timestamp::now())?;
+
+        self.store.update(*issue_id, "Reply", changes, signer)?;
+
+        Ok(())
+    }
+
+    /// Get all issues, sorted by time.
+    pub fn all(&self) -> Result<Vec<(IssueId, Issue)>, Error> {
+        let mut issues = self.store.list()?;
+        issues.sort_by_key(|(_, i)| i.timestamp);
+
+        Ok(issues)
+    }
+
+    /// Get the issue count.
+    pub fn count(&self) -> Result<usize, Error> {
+        let issues = self.store.list()?;
+
+        Ok(issues.len())
+    }
+}
+
+/// Issue events.
+mod events {
+    use super::*;
+    use automerge::{
+        transaction::{CommitOptions, Transactable},
+        ObjId,
+    };
+
+    pub fn create(
+        author: &Author,
+        title: &str,
+        description: &str,
+        timestamp: Timestamp,
+        labels: &[Label],
+    ) -> Result<Contents, TransactionError> {
+        let title = title.trim();
+        if title.is_empty() {
+            return Err(TransactionError::InvalidValue("title"));
+        }
+
+        let mut doc = Automerge::new();
+        let _issue = doc
+            .transact_with::<_, _, TransactionError, _, ()>(
+                |_| CommitOptions::default().with_message("Create issue".to_owned()),
+                |tx| {
+                    let issue = tx.put_object(ObjId::Root, "issue", ObjType::Map)?;
+
+                    tx.put(&issue, "title", title)?;
+                    tx.put(&issue, "author", author)?;
+                    tx.put(&issue, "state", State::Open)?;
+                    tx.put(&issue, "timestamp", timestamp)?;
+                    tx.put_object(&issue, "discussion", ObjType::List)?;
+
+                    let labels_id = tx.put_object(&issue, "labels", ObjType::Map)?;
+                    for label in labels {
+                        tx.put(&labels_id, label.name().trim(), true)?;
+                    }
+
+                    // Nb. The top-level comment doesn't have a `replies` field.
+                    let comment_id = tx.put_object(&issue, "comment", ObjType::Map)?;
+
+                    tx.put(&comment_id, "body", description.trim())?;
+                    tx.put(&comment_id, "author", author)?;
+                    tx.put(&comment_id, "timestamp", timestamp)?;
+                    tx.put_object(&comment_id, "reactions", ObjType::Map)?;
```

Comment:
Why is a comment automatically generated for an issue?

Reply:
It's the top-level comment, like on github issues/prs
<img width="920" alt="image" src="https://user-images.githubusercontent.com/40774/201757360-472acdac-9637-4334-a04e-26980a4701cf.png">